criu/pie/restorer.c

```diff
@@ -340,6 +359,66 @@ static inline int restore_pdeath_sig(struct thread_restore_args *ta)
 	return 0;
 }
 
+/*
+ * Map an iovec with a given file,
+ * using `sys_mmap` instead of `sys_preadv`
+ * can accelerate restore process when
+ * there are some pages we won't touch later.
+ */
+static int map_iovec(struct iovec *iovs, loff_t off, int nr, struct task_restore_args *args)
+{
+	/* Using mmap to load content as an on-demand way rather than egarly read all the content */
+	int j;
+	VmaEntry *tmp_entry;
+	unsigned long iov_size, tmp;
+	loff_t local_offset = off;
+	ssize_t local_r, r = 0;
+	for (j = 0; j < nr; j++) {
+		void *p = (void *)iovs[j].iov_base;
+		unsigned long end = (unsigned long)p + iovs[j].iov_len;
+		pr_debug("Change preadv to map of iov buffer %d, iov base %lx\n", j,
+				(unsigned long)iovs[j].iov_base);
+		/* Considering the iov buffer may cross multiple regions */
+		while ((unsigned long)p < end) {
+			/* Find the vma_entry */
+			tmp_entry = lookup(p, args);
+			iov_size = tmp_entry->end < end ? tmp_entry->end : end;
+			/* The length that we exactly want to map */
+			iov_size -= (long)p;
+
+			/*
+			 * Now check whether it can be mapped, notice that some memory areas such as `vsyscall`,
```

Comment:
vsyscall is provided by kernel and it isn't possible to change it.

Reply:
Yes, I mean vsyscall and stack can not be mapped. They still need to use `sys_pread` to load.

> notice that some memory areas such as `vsyscall`, which is not VMA_AREA_REGULAR, and stack area, whose flag has MAP_GROWSDOWN still need to use `sys_pread`.